Update chapters 5, 6 and 7 (#23)
EDUCE-UBC/MICB425: treesapp_assign_and_update.Rmd

  --output p_amoA_FunGene9.5_isolates_assign/
```

The command should take a few seconds to run and if you see a 'TreeSAPP has finished successfully.' at the end then you're good to move on.

## Analyze the classifications

### The interactive Tree of Life

Before proceeding to update the reference package we should view where the sequences were placed on the phylogeny. By just looking at where the query sequences were placed you can get a good idea of how closely related they are to the sequences in the reference package.
We will again use the interactive Tree of Life (iTOL) to visualize the JPlace file that is produced by TreeSAPP [@Letunic.2019].

You will need to copy the files to be uploaded to iTOL from the server to your local computer (i.e. laptop or desktop) using `scp`. From your local computer, run the following command to copy the 'iTOL_output' directory in the TreeSAPP outputs to your Desktop.

> ```
> scp root@<ip address>:/data/ts_tutorial/p_amoA_FunGene9.5_isolates_assign/iTOL_output ~/Desktop/
> ```

Reviewer: ````suggestion
```{bash eval = FALSE}
scp - r root@<ip address>:/data/ts_tutorial/p_amoA_FunGene9.5_isolates_assign/iTOL_output <path to Xmoa_sequences>
```
````

I believe we need the `-r` flag. Also `~` won't work with Powershell, being more general with `<path to Xmoa_sequences>` is more robust. students should have created the `Xmoa_sequences` as part of acquiring data.

> If you would like to copy the 'iTOL_output' elsewhere on your local computer, simply replace `~/Desktop/` with the path to a different directory.

Reviewer: ```suggestion
```

- Next, navigate to the page displaying the phylogeny and click-and-drag the file `XmoA_labels.txt` from your file browser into the iTOL window. This should convert the TreeSAPP identifiers (e.g. 1_XmoA) to more useful descriptions with the organism name and accession for each leaf node.

- Finally, turn on the "Phylogenetic Placements" dataset (right of the screen) and the figure should look identical to Figure \@ref(fig:XmoA-FunGene).

![(\#fig:XmoA-FunGene)The XmoA phylogeny with FunGene sequences placed onto it. The red bubbles represent sequence placements with the size of the bubble indicating the number of placements on a branch. There were 74 sequences classified, 34 of which were placed on the clade representing members of the *Gammaproteobacteria* order *Methylococcales*. Another 15 sequences were placed on the clade of *Betaproteobacteria*, represented by the genera *Nitrosospira* and *Nitrosomonas*.](images/P_AmoA_tutorial_iTOL.png)

> As you can see in the above figure, many of the sequences were placed more deeply in the phylogeny (close to the root and away from the tips of the tree). This is very different from when we classified the sequences that were used to build the reference package, if you remember from the last chapter where the sequences were placed at the tips.
> These new XmoA sequences look like they'd fill in some gaps in the tree!

Reviewer: ```suggestion
As you can see in the above figure, many of the sequences were placed more deeply in the phylogeny (close to the three's root and away from its tips). This is very different from when we classified the sequences used to build the reference package if you remember the [`treesapp create` chapter](#treesapp-create) where the sequences were placed at the tips. These new XmoA sequences look like they'd fill in some gaps in the tree, so we should use them to update it!
```

## Update the reference package with new sequences

The initial XmoA_seed reference package was pretty small with just 36 sequences (information was retrieved using `treesapp package view num_seqs -r XmoA_seed/final_outputs/XmoA_build.pkl`). Also, there were only bacterial sequences included in this version and we know there are archaeal ammonia oxidizers too! Let's see if we can expand on this seed reference package using the sequences from FunGene.

> You have already seen many of these flags and arguments before and they're used again here just to remind TreeSAPP that we want results fast. A couple of new ones are used though.\
> `treesapp update` relies on the outputs produced by `treesapp assign`, it won't take just any old FASTA file. So we'll guide it to the relevant output directory with `--treesapp_output p_amoA_FunGene9.5_isolates_assign/`. From there it will figure out what sequences were classified and what their assigned lineages were.\

Reviewer: ```suggestion
You have already seen many of these flags and arguments before, and they're used again here just to remind TreeSAPP that we want results fast. A couple of new ones are used, though. `treesapp update` won't take just any old FASTA file but relies on the outputs produced by `treesapp assign`. So we'll guide it to the relevant output directory with `--treesapp_output p_amoA_FunGene9.5_isolates_assign/`. From there, it will figure out what sequences were classified and what their assigned lineages were.

```

  --refpkg_path XmoA_seed/final_outputs/XmoA_build.pkl
```

It looks like 54 new sequences were introduced into the reference package, bringing the total to 90 sequences! And judging from the taxonomic rank summary printed during runtime both Archaea and Bacteria are now included:

```
Number of unique lineages:
	root       1
	domain     2
	phylum     5
	class      6
	order     10
	family    13
	genus     29
	species   52
```

## Tree topology after update

As before, let's view the phylogeny with the XmoA isolate sequences used to update the reference package mapped onto it. We will first need to run `treesapp assign`, then download the outputs to your local computer.

```{bash eval = FALSE}
treesapp assign \
  -i p_amoA_FunGene9.5_isolates_C65S300_uclust99.faa \
  -o XmoA_FunGene_update_isolates_assign/ \
  --refpkg_dir XmoA_FunGene_update/final_outputs/ \
  -n 4 \
  -m prot \
  --trim_align 
```

> On your local computer follow the steps for downloading the 'iTOL_outputs' directory from the server and uploading the files to iTOL as we did previously.

Reviewer: ```suggestion
On your local computer follow the steps for downloading the `iTOL_outputs` directory from the server and uploading the files to iTOL as we did previously.
```

## Annotate clades

Protein families vary in their evolutionary complexity; many, especially genes involved in scavenging and metabolizing nutrients, have often been involved in some gene duplication and/or lateral gene transfer event while others have remained sessile within their host. Moreover, closely related enzymes can vary in their activities and it is often valuable, though not necessary, to annotate such features.

If you are not familiar with the evolution of a protein family it may be helpful to find reference sequences for the activity, function, or other feature you're interested in and place them on the tree with `treesapp assign`. Then, guided by these sequences, it should be easier to colour the clades. As an example, let's place some sequences of ammonia-oxidizing Bacteria (AOB) on the phylogeny.

```{bash eval = FALSE}
treesapp assign \
  -n 4 -m prot --trim_align \
  --refpkg_dir XmoA_FunGene_update/final_outputs/ \
  --fastx_input uniprot-AmoA.fasta \
  --output AmoA_assign/
```

> Creating your own annotations involves scouring the literature to identify the phenotypes associated with each organism or taxon.For this tutorial you can use the annotations already created for the XmoA reference package.

Reviewer: ```suggestion
Creating your own annotations involves scouring the literature to identify the phenotypes associated with each organism or taxon. For this tutorial, you can use the annotations already created for the XmoA reference package.
```

```{bash eval = FALSE}
wget https://raw.githubusercontent.com/hallamlab/RefPkgs/master/Nitrogen_metabolism/Nitrification/XmoA/XmoA_taxonomy-phenotype_map.tsv
```

To create a file compatible with iTOL you will need to convert the taxonomy-phenotype mapping file with `treesapp colour`. 

```{bash eval = FALSE}
treesapp colour \
  -r XmoA_FunGene_update/final_outputs/XmoA_build.pkl \
  -o XmoA_colours/ \
  -n Function \
  --taxa_map XmoA_taxonomy-phenotype_map.tsv
```

> The two output files -- `XmoA_Function_colours_style.txt` and `XmoA_Function_colour_strip.txt` -- can be used to bring colour to the JPlace file in iTOL and also be used for adding phenotype-level annotations to the classification table with `treesapp layer`.Let's make copies of the two text files in  XmoA_colours/ into the AmoA_assign/iTOL output directory  and transfer them to desktop.

Reviewer: ```suggestion
The two output files---`XmoA_Function_colours_style.txt` and `XmoA_Function_colour_strip.txt`---can be used to bring colour to the JPlace file in iTOL and also be used for adding phenotype-level annotations to the classification table with `treesapp layer`. Let's make copies of the two text files in  XmoA_colours/ into the AmoA_assign/iTOL output directory and transfer them to `<path to Xmoa_sequences>`.
```
Em dash for a break in sentence, en dash for a range of numbers.

> ```{bash eval = FALSE}
> cp -r XmoA_colours   AmoA_assign/iTOL_output/
> scp -r  root@10.19.139.201:/data/ts_tutor/AmoA_assign/iTOL_output  < path_to_your_folder> 
> ```

Reviewer: ````suggestion
```{bash eval = FALSE}
cp -r XmoA_colours AmoA_assign/iTOL_output/
scp -r root@10.19.139.201:/data/ts_tutor/AmoA_assign/iTOL_output <path to Xmoa_sequences>
```
````

 upload the file XmoA_complete_profile.jplace. This JPlace file contains both the phylogeny and the location of the placements that will be visualized.

> Next, navigate to the page displaying the phylogeny and click-and-drag the file XmoA_labels.txt into the iTOL window. This should convert the TreeSAPP identifiers (e.g. 1_XmoA) to more useful descriptions with the organism name and accession for each leaf node.The two files XmoA_Function_colours_style.txt and XmoA_Function_colour_strip.txt  can be dragged into iTOL browser window to bring colour to the XmoA tree.Finally, turn on the "Phylogenetic Placements" dataset (right of the screen) and the figure should look like  this figure on the right.

Reviewer: ```suggestion
Next, navigate to the page displaying the phylogeny and click-and-drag the file XmoA_labels.txt into the iTOL window. This should convert the TreeSAPP identifiers (e.g. 1_XmoA) to more useful descriptions with each leaf node's organism name and accession. The two files `XmoA_Function_colours_style.txt` and `XmoA_Function_colour_strip.txt`  can be dragged into the iTOL browser window to bring colour to the XmoA tree. Finally, turn on the "Phylogenetic Placements" dataset (right of the screen), and the figure should look like this figure.
```

## Reclassify and layer annotations onto classification table

First, let's reclassify the isolate sequences from FunGene.

```{bash eval = FALSE}
treesapp assign \
  -n 4 -m prot --trim_align \
  --refpkg_dir XmoA_FunGene_update/final_outputs/ \
  --fastx_input p_amoA_FunGene9.5_isolates_C65S300_uclust99.faa \
  --output p_amoA_FunGene9.5_isolates_update_assign/
```

With the clade colours file in the tutorial directory, it is now simple to assign each of the classified query sequences to its respective functional guild with treesapp layer.

```{bash eval = FALSE}
treesapp layer \
  --colours_style XmoA_colours/XmoA_Function_colours_style.txt \
  --refpkg_dir XmoA_FunGene_update/final_outputs/ \
  --treesapp_output p_amoA_FunGene9.5_isolates_update_assign/
```

> The output of this command is p_amoA_FunGene9.5_isolates_update_assign/final_outputs/extra_annotated_marker_contig_map.tsv. It is identical to the original marker_contig_map.tsv file, but contains an extra column for the functional annotations

Reviewer: ```suggestion
The output of this command is `p_amoA_FunGene9.5_isolates_update_assign/final_outputs/extra_annotated_marker_contig_map.tsv`. It contains an extra column for the functional annotations but is otherwise identical to the original `marker_contig_map.tsv` file.
```
